README.md

```diff
@@ -1,165 +1,87 @@
+    // supported by `arrow2`
+    let struct_array= arrow_array.as_any().downcast_ref::<arrow2::array::StructArray>().unwrap();
+    assert_eq!(struct_array.len(), 3);
+
+    // deserialize back to our original vector. from_arrow() is enabled by the FromArrow trait
+    let round_trip_array: Vec<Foo> = arrow_array.from_arrow().unwrap();
+    assert_eq!(round_trip_array, original_array);
 }
 ```
 
-`FooArray::push` lays data in memory according to the arrow spec and
-can be used for all kinds of IPC, FFI, etc. supported by `arrow2`.
-
 ## Implementation details
 
-The goal is to allow the Arrow memory model to be used by an existing rust type tree and to facilitate conversions, when needed. Ideally, for performance, if the arrow memory model or specifically the API provided by the arrow2 crate exactly matches the custom type tree, then no conversions should be performed.
+The goal is to allow the Arrow memory model to be used by an existing rust type tree and to facilitate type conversions, when needed. Ideally, for performance, if the arrow memory model or specifically the API provided by the arrow2 crate exactly matches the custom type tree, then no conversions should be performed.
 
 To achieve this, the following approach is used. 
 
 - Introduce three traits, `ArrowField`, `ArrowSerialize`, and `ArrowDeserialize` that can be implemented by types that can be represented in Arrow. Implementations are provided for the built-in arrow2` types, and custom implementations can be provided for other types.
 
-- Blanket implementations are provided for types that recursively contain types that implement the above traits eg. [`Option<T>`], [`Vec<T>`], [`Vec<Option<T>>`], [`Vec<Vec<Option<T>>>`], etc. The blanked implementation needs be enabled by the `arrow_enable_vec_for_type` macro on the primitive type.
+- Blanket implementations are provided for types that recursively contain types that implement the above traits eg. [`Option<T>`], [`Vec<T>`], [`Vec<Option<T>>`], [`Vec<Vec<Option<T>>>`], etc. The blanket implementation needs be enabled by the `arrow_enable_vec_for_type` macro on the primitive type. This explicit enabling is needed since Vec<u8> is a special type in Arrow, and implementation specialization is not yet supported in rust to allow blanket implementations to coexist with more specialized implementations.
+
+- Some additional supporting traits such as `ArrowMutableArrayTryPushGeneric` are used to workaround the unavailability of features like generic associated types.
+
+### Why not serde?
+
+While, serde is the de-facto serialization framework for rust it introduces a layer of indirection that seemed unnecessary for the arrow2 framework. Specifically, arrow2 itself serves as an intermediate representation, which can be serialized/deserialized to a wide variety of formats including JSON, avro, etc, so it seemed conceptually more straightforward to have the derive functionality to have much tighter coupling with arrow2.
```

Comment:
```suggestion
While serde is the de-facto serialization framework in Rust, it introduces a layer of indirection.
Specifically, arrow2 uses Apache Arrow's in-memory columnar format, while serde is row based.
Using serde as an intermediary would lead to an extra allocation per non-primitive type per item, which is very expensive.

Arrow's in-memory format can be serialized/deserialized to a wide variety of formats including Apache Parquet, JSON, Apache Avro, Arrow IPC, and Arrow FFI specification.
```